Separate http gqtp

135yshr opened the pull request:
### 概要

Call メソッドの中でプロトコルを判別して行っていた処理をClientの作成時に分割するように処理を変更しました。
#### 変更に伴い
- GroongaClient を httpClient と gqtpClient に分割
- NewGroongaClient を NewClient に変更
### その他
- HTTP用のMockを追加
- GQTP用のMockを追加
- GQTP用の定数やリターンコードテーブルを外部から参照出来ないように変更

Commit:
Merge pull request #1 from 135yshr/patch-not_run_table_list

does not run table_list command

Commit:
Merge pull request #2 from 135yshr/patch-not_run_table_list

Patch not run table list

Commit:
add HttpClient and Client interface

httpプロトコル対応版の構造体を追加

Commit:
add testcase in skip code and new struct test code

1. 既存のテストケースをスキップするコードを追加
2. 新しく追加した HttpClient 構造体用のテストコードを追加

Commit:
mod testcase name

テストケースのメソッド名を TestHttp_Conncet_Success から TestHttp_TableList_Empty_Success に変更

Commit:
add testcase table_list command

table_list で取得できたテーブルの数が１件あることを確認するテストを追加

Commit:
add testcase column_create

column_createコマンドのテストケースを成功・失敗でそれぞれ追加

Commit:
mod JSON parse

実行結果のJSONをパースする処理を変更

パースした結果を構造体に格納後、ステータスコードを確認してエラーを返すように処理を変えた

Commit:
mod returns value of Call function

Call 関数の戻り値をGroongaResultから*GroongaResultに変更

Commit:
mod returns value of Call function

Commit:
del named return value

Commit:
del HttpClient functions

HttpClient に関連する処理を削除し、新しく http_client.go に移動

Commit:
Initial commit

HttpClientを新しいファイルに移動

Commit:
del HttpClient testcase

HttpClient に関連するテストケースを削除

Commit:
Initial commit

httpClientに関するテストケースを新しいファイルに移動

Commit:
mod NewHttpClinet to newHttpClient

newHttpClientメソッドの変更にともない、テストケースを修正

Commit:
mod NewHttpClinet to newHttpClient

公開範囲を変更
ひとつ前のコミットは、間違えて別のファイルをコミットしてしまった

Commit:
mod NewHttpClinet to newHttpClient

外部から直接インスタンスを作成できないように修正

Commit:
Refactoring Initialize functions

初期化処理を共通化

1. テスト用サーバー作成メソッド(newServer)作成
2. url.Parseのエラー処理を省略

Commit:
mod newHttpClient parameter format

Commit:
add testcase for NewClient 

NewClient関数のテストケースを追加

Commit:
add testcase unsupport protocol

http以外のプロトコルが指定されたときにNilを返すかどうかのテストケースを追加

Commit:
add NewClient function

指定されたプロトコル毎に接続するクライアント生成関数を追加

Commit:
mod remove unnecessary processing

runCommandで不要なerrチェックをしていたので削除した

Commit:
Initial commit

GQTPプロトコル用のクライアントテストファイルを新しく追加

Commit:
Initial commit

GQTPプロトコル用クライアント新規登録

Commit:
add testcase body item count

取得したBodyの数が１件（ヘッダのみ）になっているテストケースを追加

Commit:
add copy functions from goroo.go 

goroo.go から callGQTP と setResult をコピー

Commit:
add gqtp dummy server

gqtp テスト用のダミーサーバーを追加
ローカルでテストができるようにした

Commit:
del gqtpServer logic

GQTPダミーサーバーに関する処理をgqtpserver_test.goに移動

Commit:
Initial commit

GQTPプロトコル用のMockオブジェクトを新しく追加

Commit:
Initial commit

GQTPプロトコル用Mockサーバーを新しく追加
ひとつ前のコミットは、別のファイルをコミットしてしまった

Commit:
mod use channel

処理が完了するまでコネクションが切断されないように処理を修正

Commit:
mod newGqtpServer param

newGqtpServerメソッドの引数変更に伴う修正

Commit:
add testcase table count equal 1

table_list コマンドを実行したときにテーブル数が１つのときのテストを追加

Commit:
add testcase column_create

clumn_create コマンドが成功したときのテストケースを追加

Commit:
add testcase column_create return fail

column_createコマンドが失敗したときの処理を追加

Commit:
mod call function's parameter

column_create コマンドのテストなのに、table_listコマンドを設定していたので修正

Commit:
add testcase of create GQTP client 

GQTPを生成するテストケースを新しく追加

NewGroongaClientを使ったテストケースは、Skipするように処理を変更

Commit:
add gqtp protocol

GQTPプロトコルが指定されたときにgqtpClientを生成する処理を追加

Commit:
del function's

1. HTTPプロトコルとGQTPプロトコルの処理分割に伴い、共通で行っていた処理を削除
2. GQTPに関する構造体をgqtp_clientに移動

Commit:
add gqtp struct

garoo.goからGQTPプロトコルに関する構造体を移動

Commit:
del function's

HTTPプロトコルとGQTPプロトコルの分割に伴い、使わなくなったテストケースを削除した

Commit:
modify const and structure scope

外部からGQTPの定数や構造体にアクセスできないように名称を変更

Commit:
del Response struct

Response 構造体を使わなくなったので削除

Commit:
mod example

goroo.NewGroongaClientメソッドをgoroo.NewClientメソッドに変更

Commit:
add testcase select 

selectの結果が存在しないときは、body部が帰ってこないので、テストケースを追加

Commit:
add check result count

selectなどのコマンドでBody部が返ってこないケースがあったので処理を修正

Commit:
add testcase select

select コマンドでテーブルが存在しないときにBody部が返ってこないので、テストケースを追加

Commit:
mod call function

NewGroongaClientからNewClientに変更

Commit:
mod call function

NewGroongaClientをNewClientに変更

Commit:
mod function scope

メソッド名を変更
* runCommand -> run
* parseResutl -> parse

レシーバメソッドに変更
* run
* parse

Commit:
mod function scope

メソッド名を変更

 * callGQTP -> run
 * setResult -> parse

レシーバメソッドに変更

 * run
 * parse

hhatto commented:
> - GQTP用の定数やリターンコードテーブルを外部から参照出来ないように変更

これの意図はなんでしょうか？
外からリターンコード使って判定する場合を想定していたのですが、どうでしょう？

hhatto commented on goroo_test.go:
```diff
@@ -1,129 +1,67 @@
-	}
-}
+func Test_ClinetGqtp_TableList_Empty_Success(t *testing.T) {
+	body := []byte{0xc7, 0x2, 0x0, 0x0, 0x0, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0xbd, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x5b, 0x5b, 0x5b, 0x22, 0x69, 0x64, 0x22, 0x2c, 0x22, 0x55, 0x49, 0x6e, 0x74, 0x33, 0x32, 0x22, 0x5d, 0x2c, 0x5b, 0x22, 0x6e, 0x61, 0x6d, 0x65, 0x22, 0x2c, 0x22, 0x53, 0x68, 0x6f, 0x72, 0x74, 0x54, 0x65, 0x78, 0x74, 0x22, 0x5d, 0x2c, 0x5b, 0x22, 0x70, 0x61, 0x7 ...
```
適当に80文字くらいで改行を入れていただけるとありがたいです。
`gqtp_client_test.go`の方もですね。

hhatto commented:
> NewGroongaClient を NewClient に変更

これはいきなり変えるのもどうかと思うので、`// deprecated`とかのコメント入れつつ
NewGroongaClientも残しておいていただけませんか？
古いスクリプトで使ってる私が困るというw

135yshr commented:
> これの意図はなんでしょうか？
> 外からリターンコード使って判定する場合を想定していたのですが、どうでしょう？
> リターンコードが、0以外のときにerrorを返すようになっていたので、呼び出し元からは、エラーコードが参照できる必要はないかと思い、参照できないようにしました。
> 定数も同様で、Call メソッドのパラメータに使用しないので、パッケージ内から参照できるようにしてみました。

Call メソッドがとてもシンプルに作られていたので、その設計に則ってみました。

エラーコードの使用箇所

``` 修正前.go
status := uint32(resp[7]) + uint32(resp[6])<<8
if status != 0 {
    return b, fmt.Errorf("status error: [%d]%s", status, GRN_GQTP_STATUS[status])
}
```

``` 修正後.go
status := uint32(resp[7]) + uint32(resp[6])<<8
if status != 0 {
    return b, fmt.Errorf("status error: [%d]%s", status, grnReturnCode[status])
}
```

135yshr commented:
> これはいきなり変えるのもどうかと思うので、// deprecatedとかのコメント入れつつ
> NewGroongaClientも残しておいていただけませんか？
> 古いスクリプトで使ってる私が困るというw

確かにそうですよね。大変失礼致しました。
そのように修正して、もう一度プルリク送ります。

Commit:
Add NewGroongaClient in order to be compatible.

NewGroongaClientを使った既存のプログラムから使えなくなってしまうため、GroongaClientとNewGroongaClientを追加（戻し）

Commit:
Add NewGroongaClient in order to be compatible.

* NewGroongaClientを使った既存のプログラムから使えなくなってしまうため、GroongaClientとNewGroongaClientを追加（戻し）
* リファクタリング・テストデータを80文字程度で改行

Commit:
Add target version

GO言語のバージョンを1.3以外に.1.4と最新版のバージョンを追加
